### ISSUE TITLE
Issues with purged private data reconciliation

### ISSUE THREAD

**Issue by denyeart** [hyperledger/fabric#3858]
**Scenario1**
Create private data on peer1 and peer2, then stop peer2, purge private data on peer1 (block6), then restart peer2.
Upon restart of peer2, it should receive block 6 and gracefully handle the purge transaction.

**Issue1** - peer2 gets block from orderer but can’t pull private data from peer1 (presumably since peer1 has already purged it). After 1 minute of trying, peer2 commits block with missing private data:
```
2022-12-14 23:06:29.665 EST 04a6 WARN [gossip.privdata] RetrievePvtdata -> Could not fetch all 1 eligible collection private write sets for block [6]…Will commit block with missing private write sets
```
Expected behavior - When peer2 requests private data from peer1, peer1 should respond that the requested private data has been purged so that peer2 doesn’t keep trying.

Then, upon commit with missing private data, peer2 purges the private data as expected:
```
2022-12-14 23:06:29.720 EST 04d9 INFO [pvtdatastorage] deleteDataMarkedForPurge -> Processed private data purges from private data storage channel=mychannel numKeysPurged=2 numPrivateDataStoreRecordsPurged=1
```

**Issue2** - peer2 attempts reconciliation of the missing private data and succeeds:
```
2022-12-14 23:06:39.261 EST 051c DEBU [kvledger] CommitPvtDataOfOldBlocks -> [mychannel:] Committing pvtData of [1] old blocks to the pvtdatastore
2022-12-14 23:06:39.283 EST 051f INFO [gossip.privdata] reconcile -> Reconciliation cycle finished successfully. reconciled 1 private data keys from blocks range [6 - 6] channel=mychannel
```
Expected behavior: peer2 should not attempt to reconcile the missing private data in the first place, because the private data has already been purged. Shouldn’t the data be removed from the missing data queue when it got purged? Potentially this is happening because the purge logic runs prior to the commit logic, so the missing data entry gets added to the missing data queue but doesn’t get purged from the missing data queue (similar to issue #3828, so perhaps the same fix would apply).

Additionally, was the private data really reconciled per the log messages? Peer1 shouldn’t have the purged data anymore and therefore peer2 should have no way to get it. So why do we get messages on peer2 “Committing pvtData of [1] old blocks to the pvtdatastore” and “reconciled 1 private data keys from blocks range [6 - 6]”?
Peer1 log seems to show that it actually found the private data and returned it to peer2:
```
2022-12-14 23:51:06.272 EST 12f4 DEBU [gossip.privdata] filterNotEligible -> Found 1 for TxID  , collection assetCollection for 127.0.0.1:7055 channel=mychannel
2022-12-14 23:51:06.272 EST 12fc DEBU [gossip.privdata] func1 -> Returning 127.0.0.1:7055 1 elements channel=mychannel
```
But perhaps these messages are misleading... perhaps peer1 did respond to peer2 correctly that the private data is missing, and peer2 honored it by removing the entry from the missing data queue.

**Scenario2**
With peer2 down, create (block 10) and then purge (block 11) private data on peer1.
Upon restart of peer2, it should receive block 10 and block 11, and gracefully handle the create and purge transaction.

This issue starts similar to the prior issues… peer2 attempts to retrieve the private data of the ‘create’ block 10 for one minute and then gives up:
```
2022-12-15 00:00:05.138 EST 0ba1 WARN [gossip.privdata] RetrievePvtdata -> Could not fetch all 1 eligible collection private write sets for block [10]…Will commit block with missing private write sets
```
**Issue3** - when peer2 attempts to reconcile the missing private data, this time it enters a reconciliation infinite loop with peer2 constantly requesting the private data from peer1:
```
2022-12-15 00:00:17.325 EST ab19 WARN [kvledger] logSkippedCollection -> Failed to reconcile pvtdata for chaincode=private collection=assetCollection blocknum=10 txnum=0 reason="Only subset of expected keys present"
2022-12-15 00:00:17.328 EST ab48 WARN [kvledger] logSkippedCollection -> Failed to reconcile pvtdata for chaincode=private collection=assetCollection blocknum=10 txnum=0 reason="Only subset of expected keys present"
2022-12-15 00:00:17.331 EST ab77 WARN [kvledger] logSkippedCollection -> Failed to reconcile pvtdata for chaincode=private collection=assetCollection blocknum=10 txnum=0 reason="Only subset of expected keys present"
```

We may want to address Issue3 first, since we may not be able to reproduce it after Issue1 and Issue2 are resolved.

**Comment by denyeart** [comment]
Issue1 will be addressed in this issue.
Issue2 is not an issue.
Issue3 was resolved in PR #3854

Here's logs to help with Issue1.

peer3 was down while data was purged on peer1 in block 9. The logs demonstrate peer3 starting and attempting to pull the private data from peer1 for 60 seconds before giving up.

[requesting_peer3.txt](https://github.com/hyperledger/fabric/files/10242726/requesting_peer3.txt)
[replying_peer1.txt](https://github.com/hyperledger/fabric/files/10242728/replying_peer1.txt)

Extracting the main INFO statements here for context:

```
2022-12-15 23:31:45.904 EST 0115 INFO [gossip.privdata] StoreBlock -> Received block [9] from buffer channel=mychannel
2022-12-15 23:31:45.907 EST 0164 INFO [committer.txvalidator] Validate -> [mychannel] Validated block [9] in 3ms

<60 seconds of retrieve attempts>

2022-12-15 23:32:46.306 EST 0b95 WARN [gossip.privdata] RetrievePvtdata -> Could not fetch all 1 eligible collection private write sets for block [9] (0 from local cache, 0 from transient store, 0 from other peers). Will commit block with missing private write sets:[txID: 8088ec18478c82e1b3144aeeb24678de3779f021a1baa486b32b010c03fb44ad, seq: 0, namespace: private, collection: assetCollection, hash: 4650cb9dc89a557bcd4efd0164bef7eef7147fdbe95fdf2230b0504f9468f08a
] channel=mychannel
2022-12-15 23:32:46.340 EST 0bc8 INFO [pvtdatastorage] deleteDataMarkedForPurge -> Processed private data purges from private data storage channel=mychannel numKeysPurged=2 numPrivateDataStoreRecordsPurged=1
2022-12-15 23:32:46.448 EST 0be4 INFO [kvledger] commit -> [mychannel] Committed block [9] with 1 transaction(s) in 141ms (state_validation=6ms block_and_pvtdata_commit=91ms state_commit=22ms) commitHash=[2173a46924c1b4326f42f767aa8e19ce1b9cf070c43eea26fd3b6b5de468fbd3]
```

**Comment by denyeart** [comment]
Issue 1 resolved in https://github.com/hyperledger/fabric/pull/3907
